core/lib/zksync_core/src/api_server/web3/backend_jsonrpsee/batch_limiter_middleware.rs

```diff
@@ -0,0 +1,96 @@
+use std::{num::NonZeroU32, sync::Arc};
+
+use governor::{
+    clock::DefaultClock,
+    middleware::NoOpMiddleware,
+    state::{InMemoryState, NotKeyed},
+    Quota, RateLimiter,
+};
+use vise::{Buckets, Counter, EncodeLabelSet, EncodeLabelValue, Family, Histogram, Metrics};
+use zksync_web3_decl::jsonrpsee::{
+    server::middleware::rpc::{layer::ResponseFuture, RpcServiceT},
+    types::{error::ErrorCode, ErrorObject, Request},
+    MethodResponse,
+};
+
+use crate::api_server::web3::metrics::API_METRICS;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, EncodeLabelValue, EncodeLabelSet)]
+#[metrics(label = "transport", rename_all = "snake_case")]
+pub(crate) enum Transport {
+    Ws,
+}
+
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "api_jsonrpc_backend_batch")]
+struct LimitMiddlewareMetrics {
+    /// Number of rate-limited requests.
+    rate_limited: Family<Transport, Counter>,
+    /// Size of batch requests.
+    #[metrics(buckets = Buckets::exponential(1.0..=512.0, 2.0))]
+    size: Family<Transport, Histogram<usize>>,
+    /// Number of requests rejected by the limiter.
+    rejected: Family<Transport, Counter>,
+}
+
+#[vise::register]
+static METRICS: vise::Global<LimitMiddlewareMetrics> = vise::Global::new();
+pub struct LimitMiddleware<S> {
+    inner: S,
+    rate_limiter: Option<Arc<RateLimiter<NotKeyed, InMemoryState, DefaultClock, NoOpMiddleware>>>,
+    transport: Transport,
+}
+
+impl<S> LimitMiddleware<S> {
+    pub(crate) fn new(inner: S, requests_per_minute_limit: Option<u32>) -> Self {
+        API_METRICS.ws_open_sessions.inc_by(1);
+
+        Self {
+            inner,
+            rate_limiter: requests_per_minute_limit.map(|limit| {
+                Arc::new(RateLimiter::direct(Quota::per_minute(
+                    NonZeroU32::new(limit).expect("requests per minute must be > 0; qed"),
+                )))
+            }),
+            transport: Transport::Ws,
+        }
+    }
+}
+
+impl<S> Drop for LimitMiddleware<S> {
+    fn drop(&mut self) {
+        API_METRICS.ws_open_sessions.dec_by(1);
+    }
+}
+
+impl<'a, S> RpcServiceT<'a> for LimitMiddleware<S>
+where
+    S: Send + Clone + Sync + RpcServiceT<'a>,
+{
+    type Future = ResponseFuture<S::Future>;
+
+    fn call(&self, request: Request<'a>) -> Self::Future {
+        if let Some(ref rate_limiter) = self.rate_limiter {
```

Comment:
Bikeshedding: IMO, using `ref` / `ref mut` is really old-school; in most cases, it can be avoided by borrowing the matched data (i.e., `.. = &self.rate_limiter` in this case).

Reply:
changed to new school